merkle/log_verifier.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+// RootMismatchError occurs when an inclusion proof fails.
+type RootMismatchError struct {
+	ExpectedRoot   []byte
+	CalculatedRoot []byte
+}
+
+func (e RootMismatchError) Error() string {
+	return fmt.Sprintf("calculated root:\n%v\n does not match expected root:\n%v", e.CalculatedRoot, e.ExpectedRoot)
+}
+
+// LogVerifier verifies inclusion and consistency proofs for append only logs.
+type LogVerifier struct {
+	hasher TreeHasher
+}
+
+// NewLogVerifier returns a new LogVerifier for a tree.
+func NewLogVerifier(hasher TreeHasher) LogVerifier {
+	return LogVerifier{
+		hasher: hasher,
+	}
+}
+
+// VerifyInclusionProof verifies the correctness of the proof given the passed in information about the tree and leaf.
+func (m LogVerifier) VerifyInclusionProof(leafIndex, treeSize int64, proof [][]byte, root []byte, leaf []byte) error {
+	calcRoot, err := m.RootFromInclusionProof(leafIndex, treeSize, proof, leaf)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(calcRoot, root) {
+		return RootMismatchError{
+			CalculatedRoot: calcRoot,
+			ExpectedRoot:   root,
+		}
+	}
+	return nil
+}
+
+// RootFromInclusionProof calculates the expected tree root given the proof and leaf.
+// leafIndex starts at 0. treeSize starts at 1.
```

Comment:
Reword the treeSize starts at 1 comment to avoid slight potential confusion. The supplied treeSize must be > 1, trees don't start at size one. Also maybe say what the caller is expected supply in proof and leaf slices.